File: backend/app/schemas/ingredient.py
"""Ingredient request/response schemas."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

UnitType = Literal["g", "kg", "cl", "l", "piece"]


class IngredientCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    unit: UnitType
    current_price: float | None = None
    supplier_name: str | None = None


class IngredientUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    unit: UnitType | None = None
    current_price: float | None = None
    supplier_name: str | None = None


class IngredientResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    unit: str
    current_price: float | None
    supplier_name: str | None
    last_updated: datetime | None
    created_at: datetime


class IngredientListResponse(BaseModel):
    items: list[IngredientResponse]
    total: int


# --- Price history ---

class PriceHistoryEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    price: float
    date: date
    invoice_id: int | None = None
    supplier_name: str | None = None
    created_at: datetime


class PriceHistoryResponse(BaseModel):
    ingredient_name: str
    current_price: float | None
    history: list[PriceHistoryEntry]

File: frontend/src/types/ingredient.ts
// Mirror of backend/app/schemas/ingredient.py

export type UnitType = 'g' | 'kg' | 'cl' | 'l' | 'piece';

export interface Ingredient {
  id: number;
  name: string;
  unit: UnitType;
  current_price: number | null;
  supplier_name: string | null;
  category: string | null;
  last_updated: string | null;
  created_at: string;
}

export interface IngredientListResponse {
  items: Ingredient[];
  total: number;
}

export interface IngredientCreateRequest {
  name: string;
  unit: UnitType;
  current_price?: number | null;
  supplier_name?: string | null;
  category?: string | null;
}

export interface IngredientUpdateRequest {
  name?: string;
  unit?: UnitType;
  current_price?: number | null;
  supplier_name?: string | null;
  category?: string | null;
}

export interface PriceHistoryEntry {
  price: number;
  date: string;
  invoice_id: number | null;
  supplier_name: string | null;
  created_at: string;
}

export interface PriceHistoryResponse {
  ingredient_name: string;
  current_price: number | null;
  history: PriceHistoryEntry[];
}

export interface IngredientRecipeItem {
  recipe_id: number;
  recipe_name: string;
  category: string | null;
  quantity: number;
  unit: string;
}
